R/data.table.R

```diff
@@ -1739,21 +1739,27 @@ replace_dot_alias = function(e) {
       } else {
         # Apply GForce
         .gforce_ok = function(q) {
+          noCall_noVars = function(expr) (!is.call(expr) || length(all.vars(expr, max.names=1L))==0) && !dotN(expr)
           if (dotN(q)) return(TRUE) # For #334
           # run GForce for simple f(x) calls and f(x, na.rm = TRUE)-like calls where x is a column of .SD
           # is.symbol() is for #1369, #1974 and #2949
           if (!(is.call(q) && is.symbol(q[[1L]]) && is.symbol(q[[2L]]) && (q1 <- q[[1L]]) %chin% gfuns)) return(FALSE)
           if (!(q2 <- q[[2L]]) %chin% names(SDenv$.SDall) && q2 != ".I") return(FALSE)  # 875
-          if ((length(q)==2L || (!is.null(names(q)) && startsWith(names(q)[3L], "na")))) return(TRUE)
+          if (length(q)==2L || (!is.null(names(q)) && startsWith(names(q)[3L], "na") && noCall_noVars(q[[3L]]))) return(TRUE)
           #                       ^^ base::startWith errors on NULL unfortunately
           if (length(q)>=2L && q[[1L]] == "shift") {
             q_named = match.call(shift, q)
-            if (!is.call(q_named[["fill"]]) && is.null(q_named[["give.names"]])) return(TRUE)
+            if (noCall_noVars(q_named[["n"]]) &&
```

Comment:
personally, I would prefer moving this logic into a helper function like

```r
.gshift_ok <- function(q) {
  q <- match.call(shift, q)
  ...
}
```

It will reduce the indentation, isolate the logic, make it easier to test, etc.

Then we can write

```r
.gforce_ok <- function(q) {
  # handle !length(q) || !is.call(q) case
  switch(as.character(q[[1L]]),
    shift = .gshift_ok(...),
    head = , tail = .ghead_ok(...),
    ...
}
```

WDYT

Reply:
Good idea. I guess ideally we would also evaluate the arguments in `gforce_ok` if it returns true.